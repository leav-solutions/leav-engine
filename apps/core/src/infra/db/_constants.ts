// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export const ARANGODB_WRITEWRITE_CONFLICT_CODE = 1200;
export const ARANGODB_HTTP_CONFLICT_CODE = 409;
