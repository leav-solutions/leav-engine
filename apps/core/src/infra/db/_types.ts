// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
