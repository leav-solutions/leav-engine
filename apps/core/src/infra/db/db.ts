// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IConfig} from '_types/config';

interface IDeps {
    config?: IConfig;
}

const _getDbConnection = (config: IConfig): Database => {
    return new Database({
        url: config.db.url
    });
};

export default function (deps: IDeps): Database {
    const db = _getDbConnection(deps.config);

    return db.database(deps.config.db.name);
}

export const initDb = async (config: IConfig) => {
    const db = _getDbConnection(config);

    const databases = await db.listDatabases();
    const dbExists = databases.reduce((exists, d) => exists || d === config.db.name, false);

    if (!dbExists) {
        await db.createDatabase(config.db.name);
    }

    db.close();
};
