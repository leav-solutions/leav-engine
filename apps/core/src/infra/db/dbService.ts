// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {getCallStack} from '@leav/utils';
import {Database} from 'arangojs';
import {isAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {createHash} from 'crypto';
import {IUtils} from 'utils/utils';
import {IConfig} from '_types/config';
import {IDbProfiler} from '_types/dbProfiler';
import {IDbDocument, IExecute, IExecuteWithCount} from './_types';

const MAX_ATTEMPTS = 10;

export interface IDbService {
    db?: Database;

    /**
     * Execute an AQL query
     * If withTotalCount is set to true, return an object with totalCount and results. Otherwise, just return
     * results straight from DB.
     *
     * @param query
     * @param withTotalCount
     * @param attempts Used when we have to retry a query after a write-write conflict
     * @throws If query fails or we still have a conflict after all attempts
     */
    execute?<T extends IExecuteWithCount<unknown> | unknown[] = IDbDocument[]>(params: IExecute): Promise<T>;

    /**
     * Create a new collection in database
     *
     * @param name Collection name
     * @param type Document or edge collection?
     * @throws If collection already exists
     */
    createCollection?(name: string, type?: collectionTypes): Promise<void> | null;

    /**
     * Delete a collection in database
     *
     * @param name Collection name
     * @param type Document or edge collection?
     * @throws If collection already exists
     */
    dropCollection?(name: string, type?: collectionTypes): Promise<void> | null;

    /**
     * Check if collection already exists
     *
     * @param name
     */
    collectionExists?(name: string): Promise<boolean>;
}

export enum collectionTypes {
    DOCUMENT = 'document',
    EDGE = 'edge'
}

interface IDeps {
    'core.infra.db'?: Database;
    'core.utils'?: IUtils;
    config?: IConfig;
}

export default function ({
    'core.infra.db': db = null,
    'core.utils': utils = null,
    config = null
}: IDeps = {}): IDbService {
    const collectionExists = async function (name: string): Promise<boolean> {
        const collections = await db.listCollections();

        return collections.reduce((exists, c) => exists || c.name === name, false);
    };

    const _sleep = (ms: number): Promise<void> => {
        return new Promise(resolve => setTimeout(resolve, ms));
    };

    return {
        db,
        async execute<T extends IExecuteWithCount<any> | any[] = any[]>({
            query,
            ctx,
            withTotalCount = false,
            attempts = 0
        }: IExecute): Promise<T> {
            try {
                if (config.dbProfiler.enable) {
                    const dbProfiler: IDbProfiler = ctx.dbProfiler ?? {
                        totalCount: 0,
                        uniqueQueriesCount: 0,
                        queries: {}
                    };

                    dbProfiler.totalCount = (dbProfiler.totalCount ?? 0) + 1;

                    // Generate a hash from the query to be able
                    // to group identical queries (exact same query with exact same params)
                    const queryKey = createHash('md5').update(JSON.stringify(query)).digest('base64');

                    if (!dbProfiler.queries) {
                        dbProfiler.queries = {};
                    }

                    const callStack = JSON.stringify(getCallStack(10));

                    dbProfiler.queries[queryKey] = {
                        count: (dbProfiler.queries?.[queryKey]?.count ?? 0) + 1,
                        callers: (dbProfiler.queries?.[queryKey]?.callers ?? new Set()).add(callStack),
                        query
                    };

                    dbProfiler.uniqueQueriesCount = Object.keys(dbProfiler.queries).length;

                    ctx.dbProfiler = dbProfiler;
                }

                // Convert query to AqlQuery if we have a simple query to match query() types
                const queryToRun = isAqlQuery(query)
                    ? {...query}
                    : {
                          query,
                          bindVars: {}
                      };

                const queryOptions = withTotalCount ? {count: true, options: {fullCount: true}} : {};

                const cursor = await db.query(queryToRun, queryOptions);

                const results = await cursor.all();
                return withTotalCount ? {totalCount: cursor.extra.stats.fullCount, results} : results;
            } catch (e) {
                // Handle write-write conflicts: we try the query again with a growing delay between trials.
                // If we reach maximum attempts and still no success, stop it and throw the exception
                // error 1200 === conflict
                if (e.isArangoError && e.errorNum === 1200 && attempts < MAX_ATTEMPTS) {
                    const timeToWait = 2 ** attempts;
                    await _sleep(timeToWait);
                    return this.execute({query, ctx, withTotalCount, attempts: attempts + 1});
                }

                e.message += `\nQuery was: ${JSON.stringify(query).replace(/\\n/g, ' ')}`;
                e.query = query;

                // Response contains circular references which can cause an error when converted to JSON
                // later on. It doesn't contains useful information anyway, so throw it away.
                delete e.response;

                utils.rethrow(e);
            }
        },
        async createCollection(name: string, type = collectionTypes.DOCUMENT): Promise<void> {
            if (await collectionExists(name)) {
                throw new Error(`Collection ${name} already exists`);
            }

            if (type === collectionTypes.EDGE) {
                const collection = db.edgeCollection(name);
                await collection.create();
            } else {
                const collection = db.collection(name);
                await collection.create();
            }
        },
        async dropCollection(name: string, type = collectionTypes.DOCUMENT): Promise<void> {
            if (!(await collectionExists(name))) {
                throw new Error(`Collection ${name} does not exist`);
            }

            const collection = type === collectionTypes.EDGE ? db.edgeCollection(name) : db.collection(name);
            await collection.drop();
        },
        collectionExists
    };
}
