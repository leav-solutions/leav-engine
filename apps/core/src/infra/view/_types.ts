// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IList} from '_types/list';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IQueryInfos} from '../../_types/queryInfos';

export const VIEWS_COLLECTION_NAME = 'core_views';

export interface IViewRepo {
    createView(view: IView, ctx: IQueryInfos): Promise<IView>;
    updateView(view: IView, ctx: IQueryInfos): Promise<IView>;
    getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>>;
    deleteView(viewId: string, ctx: IQueryInfos): Promise<IView>;
}

export interface IViewRepoDeps {
    'core.infra.db.dbService': IDbService;
    'core.infra.db.dbUtils': IDbUtils;
}
