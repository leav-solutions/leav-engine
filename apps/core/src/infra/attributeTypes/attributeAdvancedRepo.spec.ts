// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql, Database} from 'arangojs';
import {IDbUtils} from 'infra/db/dbUtils';
import {IQueryInfos} from '_types/queryInfos';
import {AttributeTypes} from '../../_types/attribute';
import {AttributeCondition} from '../../_types/record';
import {mockAttrAdvVersionableSimple} from '../../__tests__/mocks/attribute';
import attributeAdvancedRepo from './attributeAdvancedRepo';

describe('AttributeStandardRepo', () => {
    const mockAttribute = {
        id: 'test_attr',
        type: AttributeTypes.ADVANCED,
        multiple_values: true
    };

    const mockDbUtils: Mockify<IDbUtils> = {
        convertValueVersionToDb: jest.fn().mockReturnValue({my_tree: 'test_lib/1'}),
        convertValueVersionFromDb: jest.fn().mockReturnValue({
            my_tree: {
                id: '1',
                library: 'test_lib'
            }
        })
    };
    const ctx: IQueryInfos = {
        userId: '0',
        queryId: 'attributeAdvancedRepoTest'
    };

    describe('createValue', () => {
        test('Should create a new advanced value', async function () {
            const createdValueData = {
                _id: 'core_values/987654',
                _rev: '_WSywvyC--_',
                _key: 987654,
                value: 'test_val'
            };

            const createdEdgeData = {
                _id: 'core_edge_values_links/222435651',
                _rev: '_WSywvyC--_',
                _from: 'test_lib/12345',
                _to: 'core_values/987654',
                _key: 978654321,
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0',
                metadata: {my_attribute: 'metadata value'}
            };

            const newValueData = {
                id_value: 987654,
                value: 'test_val',
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                created_by: '0',
                modified_by: '0',
                metadata: {my_attribute: 'metadata value'}
            };

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromiseMultiple([[createdValueData], [createdEdgeData]])
            };

            const attrRepo = attributeAdvancedRepo({'core.infra.db.dbService': mockDbServ});

            const createdVal = await attrRepo.createValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 'test val',
                    modified_at: 400999999,
                    created_at: 400999999,
                    metadata: {my_attribute: 'metadata value'}
                },
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(2);

            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/INSERT/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(typeof mockDbServ.execute.mock.calls[1][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatch(/INSERT/);
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[1][0].query.bindVars).toMatchSnapshot();

            expect(createdVal).toMatchObject(newValueData);
        });

        test('Should save version on value', async function () {
            const createdValueData = {
                _id: 'core_values/987654',
                _rev: '_WSywvyC--_',
                _key: 987654,
                value: 'test_val',
                version: {
                    my_tree: 'test_lib/1'
                }
            };

            const createdEdgeData = {
                _id: 'core_edge_values_links/222435651',
                _rev: '_WSywvyC--_',
                _from: 'test_lib/12345',
                _to: 'core_values/987654',
                _key: 978654321,
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0',
                version: {
                    my_tree: 'test_lib/1'
                }
            };

            const newValueData = {
                id_value: 987654,
                value: 'test_val',
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0',
                version: {
                    my_tree: {
                        id: '1',
                        library: 'test_lib'
                    }
                }
            };

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromiseMultiple([[createdValueData], [createdEdgeData]])
            };

            const attrRepo = attributeAdvancedRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const createdVal = await attrRepo.createValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    value: 'test val',
                    modified_at: 400999999,
                    created_at: 400999999,
                    version: {
                        my_tree: {
                            id: '1',
                            library: 'test_lib'
                        }
                    }
                },
                ctx
            });

            expect(typeof mockDbServ.execute.mock.calls[1][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatch(/INSERT/);
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[1][0].query.bindVars).toMatchSnapshot();

            expect(createdVal).toMatchObject(newValueData);
        });
    });

    describe('updateValue', () => {
        test('Should update an advanced value', async function () {
            const savedValueData = {
                _id: 'core_values/987654',
                _rev: '_WSywvyC--_',
                _key: 987654,
                value: 'test_val'
            };

            const savedEdgeData = {
                _id: 'core_edge_values_links/222435651',
                _rev: '_WSywvyC--_',
                _from: 'test_lib/12345',
                _to: 'core_values/987654',
                _key: 978654321,
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0',
                metadata: {my_attribute: 'metadata value'}
            };

            const valueData = {
                id_value: 987654,
                value: 'test_val',
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0',
                metadata: {my_attribute: 'metadata value'}
            };

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromiseMultiple([[savedValueData], [savedEdgeData]])
            };

            const attrRepo = attributeAdvancedRepo({'core.infra.db.dbService': mockDbServ});

            const savedVal = await attrRepo.updateValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    id_value: '987654',
                    value: 'test val',
                    modified_at: 500999999,
                    metadata: {my_attribute: 'metadata value'}
                },
                ctx
            });

            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/UPDATE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(typeof mockDbServ.execute.mock.calls[1][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatch(/UPDATE/);
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[1][0].query.bindVars).toMatchSnapshot();

            expect(savedVal).toMatchObject(valueData);
        });

        test('Should update value version', async function () {
            const savedValueData = {
                _id: 'core_values/987654',
                _rev: '_WSywvyC--_',
                _key: 987654,
                value: 'test_val',
                version: {
                    my_tree: 'test_lib/1'
                }
            };

            const savedEdgeData = {
                _id: 'core_edge_values_links/222435651',
                _rev: '_WSywvyC--_',
                _from: 'test_lib/12345',
                _to: 'core_values/987654',
                _key: 978654321,
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0',
                version: {
                    my_tree: 'test_lib/1'
                }
            };

            const valueData = {
                id_value: 987654,
                value: 'test_val',
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0',
                version: {
                    my_tree: {
                        id: '1',
                        library: 'test_lib'
                    }
                }
            };

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromiseMultiple([[savedValueData], [savedEdgeData]])
            };

            const attrRepo = attributeAdvancedRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const savedVal = await attrRepo.updateValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    id_value: '987654',
                    value: 'test val',
                    modified_at: 500999999,
                    version: {
                        my_tree: {
                            id: '1',
                            library: 'test_lib'
                        }
                    }
                },
                ctx
            });

            expect(savedVal).toMatchObject(valueData);
        });
    });

    describe('deleteValue', () => {
        test('Should delete a value', async function () {
            const deletedValueData = {
                _id: 'core_values/123456789',
                _rev: '_WSywvyC--_',
                _key: 123456789,
                value: 'test_val'
            };

            const deletedEdgeData = {
                _id: 'core_edge_values_links/222435651',
                _rev: '_WSywvyC--_',
                _from: 'test_lib/12345',
                _to: 'core_values/987654',
                _key: 978654321,
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999,
                modified_by: '0',
                created_by: '0'
            };

            const oldValueData = {
                id_value: 123456789,
                attribute: 'test_attr',
                modified_at: 400999999,
                created_at: 400999999
            };

            const mockDbCollec = {
                remove: global.__mockPromise(deletedValueData)
            };

            const mockDbEdgeCollec = {
                removeByExample: global.__mockPromise(deletedEdgeData)
            };

            const mockDb = {
                collection: jest.fn().mockReturnValue(mockDbCollec),
                edgeCollection: jest.fn().mockReturnValue(mockDbEdgeCollec)
            };

            const mockDbServ = {db: (mockDb as unknown) as Database};

            const attrRepo = attributeAdvancedRepo({'core.infra.db.dbService': mockDbServ});

            const deletedVal = await attrRepo.deleteValue({
                library: 'test_lib',
                recordId: '12345',
                attribute: mockAttribute,
                value: {
                    id_value: '123456789',
                    value: 'test val',
                    modified_at: 400999999,
                    created_at: 400999999,
                    modified_by: '0',
                    created_by: '0'
                },
                ctx
            });

            expect(mockDbCollec.remove.mock.calls.length).toBe(1);
            expect(mockDbCollec.remove).toBeCalledWith({_key: '123456789'});

            expect(mockDbEdgeCollec.removeByExample.mock.calls.length).toBe(1);
            expect(mockDbEdgeCollec.removeByExample).toBeCalledWith({
                _from: 'test_lib/12345',
                _to: 'core_values/123456789'
            });

            expect(deletedVal).toMatchObject(oldValueData);
        });
    });

    describe('getValueByID', () => {
        test('Should return value', async function () {
            const lookupValueRes = [
                {
                    _key: 987654,
                    value: 'test val'
                }
            ];

            const edgeRes = [
                {
                    _from: 'test_lib/987654',
                    _to: 'core_values/987654',
                    modified_at: 99999,
                    created_at: 99999,
                    modified_by: '0',
                    created_by: '0',
                    attribute: 'test_attr'
                }
            ];

            const mockDbCollec = {
                lookupByKeys: global.__mockPromise(lookupValueRes)
            };

            const mockDbEdgeCollec = {
                inEdges: global.__mockPromise(edgeRes)
            };

            const mockDb = {
                collection: jest.fn().mockReturnValue(mockDbCollec),
                edgeCollection: jest.fn().mockReturnValue(mockDbEdgeCollec)
            };

            const mockDbServ = {db: (mockDb as unknown) as Database};

            const attrRepo = attributeAdvancedRepo({'core.infra.db.dbService': mockDbServ});

            const value = await attrRepo.getValueById({
                library: 'test_lib',
                recordId: '987654',
                attribute: mockAttribute,
                valueId: '132465',
                ctx
            });

            expect(mockDbCollec.lookupByKeys.mock.calls.length).toBe(1);
            expect(mockDbEdgeCollec.inEdges.mock.calls.length).toBe(1);
            expect(value).toMatchObject({
                id_value: 987654,
                value: 'test val',
                modified_at: 99999,
                created_at: 99999,
                modified_by: '0',
                created_by: '0',
                attribute: 'test_attr'
            });
        });

        test("Should return null if value doesn't exists", async function () {
            const mockDbCollec = {
                lookupByKeys: global.__mockPromise([])
            };

            const mockDbEdgeCollec = {
                inEdges: global.__mockPromise()
            };

            const mockDb = {
                collection: jest.fn().mockReturnValue(mockDbCollec),
                edgeCollection: jest.fn().mockReturnValue(mockDbEdgeCollec)
            };

            const mockDbServ = {db: (mockDb as unknown) as Database};

            const attrRepo = attributeAdvancedRepo({'core.infra.db.dbService': mockDbServ});

            const value = await attrRepo.getValueById({
                library: 'test_lib',
                recordId: '987654',
                attribute: mockAttribute,
                valueId: '132465',
                ctx
            });

            expect(mockDbCollec.lookupByKeys.mock.calls.length).toBe(1);
            expect(mockDbEdgeCollec.inEdges.mock.calls.length).toBe(0);
            expect(value).toBeNull();
        });
    });

    describe('getValues', () => {
        const traversalRes = [
            {
                value: {
                    _key: 987654,
                    value: 'test val'
                },
                edge: {
                    _from: 'test_lib/123456',
                    _to: 'core_values/987654',
                    attribute: 'test_attr',
                    modified_at: 99999,
                    created_at: 99999,
                    modified_by: '0',
                    created_by: '0',
                    metadata: {my_attribute: 'metadata value'}
                }
            },
            {
                value: {
                    _key: 987655,
                    value: 'test val2'
                },
                edge: {
                    _from: 'test_lib/123456',
                    _to: 'core_values/987655',
                    attribute: 'test_attr',
                    modified_at: 99999,
                    created_at: 99999,
                    modified_by: '0',
                    created_by: '0',
                    metadata: {my_attribute: 'metadata value'}
                }
            }
        ];

        test('Should return values for advanced attribute', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalRes)
            };

            const attrRepo = attributeAdvancedRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttribute,
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(values.length).toBe(2);
            expect(values[0]).toMatchObject({
                id_value: 987654,
                value: 'test val',
                attribute: 'test_attr',
                modified_at: 99999,
                created_at: 99999,
                modified_by: '0',
                created_by: '0',
                metadata: {my_attribute: 'metadata value'}
            });
        });

        test('Should return only first value if not multiple values', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([traversalRes[0]])
            };

            const attrRepo = attributeAdvancedRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const mockAttrNotMultival = {
                ...mockAttribute,
                multiple_values: false
            };

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttrNotMultival,
                ctx
            });

            expect(values.length).toBe(1);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch('LIMIT 1');
            expect(values[0]).toMatchObject({
                id_value: 987654,
                value: 'test val',
                attribute: 'test_attr',
                modified_at: 99999,
                created_at: 99999,
                modified_by: '0',
                created_by: '0'
            });
        });

        test('Should return values filtered by version', async function () {
            const traversalResWithVers = [
                {
                    value: {
                        _key: 987654,
                        value: 'test val'
                    },
                    edge: {
                        _from: 'test_lib/123456',
                        _to: 'core_values/987654',
                        attribute: 'test_attr',
                        modified_at: 99999,
                        created_at: 99999,
                        modified_by: '0',
                        created_by: '0',
                        version: {
                            my_tree: 'my_lib/1345'
                        }
                    }
                }
            ];

            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalResWithVers)
            };

            const attrRepo = attributeAdvancedRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttrAdvVersionableSimple,
                forceGetAllValues: false,
                options: {
                    version: {my_tree: {library: 'my_lib', id: '1345'}}
                },
                ctx
            });

            expect(values).toHaveLength(1);
            expect(values[0].value).toBe('test val');
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch('FILTER edge.version');
        });

        test('Should return all values if forced', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(traversalRes)
            };

            const attrRepo = attributeAdvancedRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const mockAttrNotMultival = {
                ...mockAttribute,
                multiple_values: false
            };

            const values = await attrRepo.getValues({
                library: 'test_lib',
                recordId: '123456',
                attribute: mockAttrNotMultival,
                forceGetAllValues: true,
                ctx
            });

            expect(values.length).toBe(2);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
        });
    });

    describe('filterQueryPart', () => {
        test('Should return advanced filter', () => {
            const mockDbServ = {
                db: new Database()
            };
            const attrRepo = attributeAdvancedRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.attributeTypes.helpers.getConditionPart': () => aql`rVal == ${'MyLabel'}`
            });
            const filter = attrRepo.filterQueryPart(
                [{id: 'label', type: AttributeTypes.ADVANCED, _repo: null}],
                {condition: AttributeCondition.EQUAL, value: 'MyLabel'},
                'r'
            );

            expect(filter.query).toMatch(/FILTER/);
            expect(filter).toMatchSnapshot();
        });
    });

    describe('sortQueryPart', () => {
        test('Should return advanced filter', () => {
            const mockDbServ = {
                db: new Database()
            };
            const attrRepo = attributeAdvancedRepo({'core.infra.db.dbService': mockDbServ});
            const filter = attrRepo.sortQueryPart({
                attributes: [{id: 'label', type: AttributeTypes.ADVANCED}],
                order: 'ASC'
            });

            expect(filter.query).toMatch(/^SORT/);
            expect(filter).toMatchSnapshot();
        });
    });
});
