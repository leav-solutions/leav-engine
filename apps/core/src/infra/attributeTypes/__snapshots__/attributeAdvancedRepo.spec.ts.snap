// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeStandardRepo createValue Should create a new advanced value 1`] = `
"
                    INSERT @value0
                    IN @@value1
                    RETURN NEW"
`;

exports[`AttributeStandardRepo createValue Should create a new advanced value 2`] = `
Object {
  "@value1": "core_values",
  "value0": Object {
    "value": "test val",
  },
}
`;

exports[`AttributeStandardRepo createValue Should create a new advanced value 3`] = `
"
                    INSERT @value0
                    IN @@value1
                    RETURN NEW"
`;

exports[`AttributeStandardRepo createValue Should create a new advanced value 4`] = `
Object {
  "@value1": "core_edge_values_links",
  "value0": Object {
    "_from": "test_lib/12345",
    "_to": "core_values/987654",
    "attribute": "test_attr",
    "created_at": 400999999,
    "created_by": "0",
    "metadata": Object {
      "my_attribute": "metadata value",
    },
    "modified_at": 400999999,
    "modified_by": "0",
  },
}
`;

exports[`AttributeStandardRepo createValue Should save version on value 1`] = `
"
                    INSERT @value0
                    IN @@value1
                    RETURN NEW"
`;

exports[`AttributeStandardRepo createValue Should save version on value 2`] = `
Object {
  "@value1": "core_edge_values_links",
  "value0": Object {
    "_from": "test_lib/12345",
    "_to": "core_values/987654",
    "attribute": "test_attr",
    "created_at": 400999999,
    "created_by": "0",
    "modified_at": 400999999,
    "modified_by": "0",
    "version": Object {
      "my_tree": "test_lib/1",
    },
  },
}
`;

exports[`AttributeStandardRepo filterQueryPart Should return advanced filter 1`] = `
Object {
  "_source": [Function],
  "bindVars": Object {
    "@value0": "core_edge_values_links",
    "value1": "label",
    "value2": "MyLabel",
  },
  "query": "FILTER FIRST(
                FOR rv, re IN 1 OUTBOUND r._id
                @@value0
                FILTER re.attribute == @value1 RETURN rv.value
            ) == @value2",
}
`;

exports[`AttributeStandardRepo getValues Should return all values if forced 1`] = `
"
                FOR value, edge
                IN 1 OUTBOUND @value0
                @@value1
                FILTER edge.attribute == @value2
                 
                
                RETURN {value, edge}
            "
`;

exports[`AttributeStandardRepo getValues Should return values filtered by version 1`] = `
"
                FOR value, edge
                IN 1 OUTBOUND @value0
                @@value1
                FILTER edge.attribute == @value2
                 FILTER edge.version == @value3 
                LIMIT 1
                RETURN {value, edge}
            "
`;

exports[`AttributeStandardRepo getValues Should return values for advanced attribute 1`] = `
"
                FOR value, edge
                IN 1 OUTBOUND @value0
                @@value1
                FILTER edge.attribute == @value2
                 
                
                RETURN {value, edge}
            "
`;

exports[`AttributeStandardRepo getValues Should return values for advanced attribute 2`] = `
Object {
  "@value1": "core_edge_values_links",
  "value0": "test_lib/123456",
  "value2": "test_attr",
}
`;

exports[`AttributeStandardRepo sortQueryPart Should return advanced filter 1`] = `
Object {
  "_source": [Function],
  "bindVars": Object {
    "@value0": "core_edge_values_links",
    "value1": "label",
    "value2": "ASC",
  },
  "query": "SORT FIRST(
                FOR v, e IN 1 OUTBOUND r._id
                @@value0
                FILTER e.attribute == @value1 RETURN v.value
            ) @value2",
}
`;

exports[`AttributeStandardRepo updateValue Should update an advanced value 1`] = `
"
                    UPDATE @value0
                    WITH @value1
                    IN @@value2
                    RETURN NEW"
`;

exports[`AttributeStandardRepo updateValue Should update an advanced value 2`] = `
Object {
  "@value2": "core_values",
  "value0": Object {
    "_key": "987654",
  },
  "value1": Object {
    "value": "test val",
  },
}
`;

exports[`AttributeStandardRepo updateValue Should update an advanced value 3`] = `
"
                    FOR e IN @@value0
                    FILTER e._from == @value1 AND e._to == @value2
                    UPDATE e
                        WITH @value3
                        IN @@value0
                    RETURN NEW"
`;

exports[`AttributeStandardRepo updateValue Should update an advanced value 4`] = `
Object {
  "@value0": "core_edge_values_links",
  "value1": "test_lib/12345",
  "value2": "core_values/987654",
  "value3": Object {
    "_from": "test_lib/12345",
    "_to": "core_values/987654",
    "attribute": "test_attr",
    "created_at": undefined,
    "created_by": undefined,
    "metadata": Object {
      "my_attribute": "metadata value",
    },
    "modified_at": 500999999,
    "modified_by": "0",
  },
}
`;
